### DISCUSSION TITLE
Declutter Behavior with different zIndex values

### DISCUSSION THREAD

**Discussion by AlbrechtWM** [openlayers/openlayers#16070] [General]
Before I submit this as a bug report, I wanted to hear from people about this particular topic.

For reference, I am using openLayers 9.2.4

The Issue at hand:

I have a large number of features (in this case, circles with a text label above) I am testing with (10,000).  These 10,000 features have been randomly generated within a lat/lon rectangle off of the atlantic.  The generation was done before hand, and put in a postgres database, where they are being queried, turned into geoJson and then displayed on an openLayers map.

In addition, the 10,000 features are further stratified into 1000 feature "tiers".  Tier 0 has the highest priority, Tier 9 the lowest.

All features are being added to a single vectorLayer with declutter enabled.  The features are assigned a zIndex in the styleFunction based on their Tier, with the higher priority Tiers having a corespondingly higher zIndex (note, I have tested this with multiple layers also with similar issues).

The features are color coded based on their tier, with bright green being the highest priority, and red being the lowest priority.  The following two images (seems to be capping the amount I can upload) show the progression of what's being displayed, starting at the highest zoom (farthest away) and moving progressively lower (zooming in).

![image](https://github.com/user-attachments/assets/98a110fe-82c0-4773-a70a-2fda6560902d)
![image](https://github.com/user-attachments/assets/2df6e637-b387-4b9d-9bb5-0a71b3955ff9)

The problem here, is that at the highest zoom, I would expect to see *only* bright green (highest priority, highest zIndex) features, yet some lower priority features are bubbling through.  If I zoom in, I can see that the places lower priority features were occupying at the zoom above, do have green features below, so it's not like there was a gap they were able to fill.

Is this expected behavior?  It seems like declutter *partially* incorporates the zIndex when it's running, but not fully enough to prevent lower zIndex features from "taking the place" of higher zIndex features which should be displayed at that particular zoom.

Any comments/feedback on this would be appreciated.  I am happy to provide more information as needed.

**Comment by AlbrechtWM** [comment]
A further image showing the zoom progression and the declutter behavior: 

![image](https://github.com/user-attachments/assets/7954fdd8-7d6c-4b5e-85a6-656e57912552)

**Comment by AlbrechtWM** [comment]
(I was trying to upload a very zoomed in image showing things behaving as expected, with enough space to render the features with their text, but I appear to have hit a limit on the amount of images I can upload here)

**Comment by ahocevar** [comment]
I think the behavior is fine, but without seeing your code and data, I do at least I have an explanation for what you're seeing:

Note that decluttering is determined by the bounding boxes of rendered symbols. In the case of circles, the bounding boxes are squares. Now what can happen is that when all circles with the highest zIndex have been decluttered, there may still be bounding boxes from a lower zIndex which can fit into a space that no bounding box from the highest zIndex was able to occupy because it hit other square bounding boxes around itself, and was omitted because of that.

**Reply by AlbrechtWM** [reply]
I see what you're saying re: Square hit detection for a circle.  My point is that this area is hyper saturated with points at this zoom that I just don't see how there isn't another (high priority, high zIndex, green) point that could not fill the spot that say, the dark red dot on the first image occupied.  

To illustrate, here is the map zoomed all the way out with declutter = false:

![image](https://github.com/user-attachments/assets/67b34680-08bf-4d73-a16a-0a458388928a)

You can see how many features are present here.  Also note that it's just a sea of green at this point, as the 1000 green, high zIndex features are being drawn on top.  It's only when zIndex is combined with declutter we get these other points poking through.

**Reply by AlbrechtWM** [reply]
Here is a very zoomed in view (with declutter = true) just to show you what the data looks like at when it's decluttering / rendering as I would expect it to.

![image](https://github.com/user-attachments/assets/c3009020-9742-4466-a698-07ad835619bb)

**Reply by AlbrechtWM** [reply]
Happy to provide data / code samples or any other information if it will help.  And I appreciate you taking the time to respond.

**Reply by ahocevar** [reply]
There's not a contradiction between what you're saying and what I'm saying. It's a matter of probability. And you may be underestimating the probability of no other bounding box of the same zIndex being able to fill a spot. That's because decluttering is done by simple first come first serve, per zIndex. So the first bounding box that happens to fit a free spot takes it. The spot that was finally taken by a red circle was probably one where not many green candidates were left at the end of that zIndex cycle, so a red one was able to finally take it.

When you provide your data here, this could also be shown with a side-by-side view or a decluttered and non-decluttered version of the data and some interactivity to explore the data.

**Reply by AlbrechtWM** [reply]
Ok, so speaking on the probabilty/distribution you are probably right with regards to that red point in the top right (lowest possible zIndex).  However, there is still some behavior I find strange that perhaps your could shine some light on with respects to the Declutter + zIndex algorithm.

Here is the original 10,000 point plot, with 10 tiers of zIndex and declutter on, as I showed earlier.

![image](https://github.com/user-attachments/assets/a4823de2-9847-4c3a-95df-eefef4216028)

Next, I went ahead and altered the database query to not include the final, lowest tier (the red most points).  Here is a 9000 point plot, with just 9 tiers of zIndex and declutter on.

![image](https://github.com/user-attachments/assets/b6df9c8b-f812-46ae-910d-e81f07415994)

At first glance, everything looks as expected.  The very red point in the top center disappears as you'd expect based on your assertions.  However, here is an image showing some analysis at what is happening at this maximum zoom level:

![image](https://github.com/user-attachments/assets/30a680db-911b-4258-8337-fd05fe1ce41f)

Note a couple of things:
1) Two (not just one) points have been removed from the second 9000 point plot.  Total count was 83 rendered in the first test, and 81 rendered in the second.  I would have expected only 1 point to be removed, the red point that falls within the lowest zIndex category.
2) The structure in various places has changed.  Why this confuses me, is if the algorithm is going through all the highest zIndexes, and trying to place them first, then going to the next highest zIndex, etc.  I would expect the behavior to be deterministic.  The structure changing makes it look like it's not behaving the way you describe.
3) A slightly lower tier (the greens are very similar I know, but i checked the RGB) point is replacing a formerly all green point.  Why would this occur if this space was able to be occupied in the previous run by a highest zIndex point?

Thanks, and appreciate once again you taking the time to weigh in on this.

**Reply by AlbrechtWM** [reply]
I can create a new test setup with only 5 tiers which would make the colors more visually distinct if that helps.  I can also provide a postgres backup of the data or the python script I used to generate the data for the test if you like.  Style function code also.  Let me know if you need anything.

**Reply by ahocevar** [reply]
Also this can be explained: when you remove data, the rtree of the vector source will be different, causing a different order of the features. Remeber the first come - first serve behavior? Other points will come first now in each zIndex.

**Reply by AlbrechtWM** [reply]
Certainly, however the data is presorted in this case.  The geoJson feature collection getting passed is in the exact same order, minus the last 1000 "red" (low priority, low zIndex) points which come at the end of the 10,000 features.  The data is presorted in descending order of zIndex, so why would removing the last 1000 at the end of the list cause this non-deterministic behavior with regards to how the prior 9000 points are plotted?  

Can you point me to where the code from this is handled?  Maybe if I look at it, I might get a better understanding of what it's doing.

**Reply by ahocevar** [reply]
The render order is different than the order in your source, because of spatial indexing with rbush. You can verify that by doing `source.getFeatures()`, which will return a different order than your GeoJSON source data.

If you want to maintain the original order, at the cost of a performance loss when rendering only a subset of the source's features (because of a smaller viewport), you can configure the source with `useSpatialIndex: false`.

**Reply by AlbrechtWM** [reply]
Interesting, ok the fact it's being stored in a tree would make sense why the behavior would change when the last 1000 points are no longer added (as the tree structure would change).  That makes sense, thank you.  I will give this a try with useSpatialIndex: false just to see how it looks.

**Reply by AlbrechtWM** [reply]
useSpatialIndex: false did the trick!  Thank you ahocevar!

**Comment by AlbrechtWM** [comment]
My questions regarding declutter behavior were answered.  We can now close this.
